test/e2e/functional/tests/e2e.go

```diff
@@ -354,6 +365,826 @@ var _ = Describe("Ingress Node Firewall", func() {
 		})
 	})
 
+	Context("[TCP] Rules", func() {
+		var (
+			client                               *corev1.Pod
+			server                               *corev1.Pod
+			config                               *ingressnodefwv1alpha1.IngressNodeFirewallConfig
+			inf                                  *ingressnodefwv1alpha1.IngressNodeFirewall
+			err                                  error
+			serverPodName                        = "e2e-inf-rules-server-"
+			serverListenPort                     = "80"
+			additionalServerPodName              = "e2e-inf-rules-additional-server-"
+			additionalAdditionalServerPodName    = "e2e-inf-rules-additional-additional-server-"
+			additionalAdditionalServerListenPort = "8080"
+			defaultSourcePort                    = "40000"
+			clientPodName                        = "e2e-inf-rules-client-"
+			podLabel                             = map[string]string{"e2e-inf-rules": ""}
+		)
+
+		BeforeEach(func() {
+			//setup INF config
+			config = &ingressnodefwv1alpha1.IngressNodeFirewallConfig{}
+			err := infwutils.LoadIngressNodeFirewallConfigFromFile(config, consts.IngressNodeFirewallConfigCRFile)
+			Expect(err).ShouldNot(HaveOccurred())
+			config.SetNamespace(OperatorNameSpace)
+			err = infwutils.EnsureIngressNodeFirewallConfigExists(testclient.Client, config, timeout)
+			Expect(err).ShouldNot(HaveOccurred())
+			//wait for daemonset to be rolled out
+			infDaemonSet := &appsv1.DaemonSet{}
+			err = daemonset.WaitForDaemonSetReady(testclient.Client, infDaemonSet, OperatorNameSpace,
+				"ingress-node-firewall-daemon", retryInterval, timeout)
+			Expect(err).ShouldNot(HaveOccurred())
+			//setup client pod which we always need
+			client = pods.GetNetcatClientPod(clientPodName, OperatorNameSpace, podLabel)
+			client, err = pods.EnsurePodRunning(testclient.Client, client, OperatorNameSpace, retryInterval, timeout)
+			Expect(err).ShouldNot(HaveOccurred())
+			server = pods.GetNetcatServerPod(serverPodName, OperatorNameSpace, serverListenPort, podLabel,
+				ingressnodefwv1alpha1.ProtocolTypeTCP)
+			server, err = pods.EnsurePodRunning(testclient.Client, server, OperatorNameSpace, retryInterval, timeout)
+			Expect(err).Should(BeNil())
+			inf = &ingressnodefwv1alpha1.IngressNodeFirewall{}
+			inf.SetName("e2e-inf-tcp-rules")
+			infwutils.DefineWithWorkerNodeSelector(inf)
+			infwutils.DefineWithInterface(inf, testInterface)
+			infwutils.AppendIngress(inf, fmt.Sprintf("%s/32", client.Status.PodIP),
+				infwutils.GetTCPRule(1, serverListenPort, ingressnodefwv1alpha1.IngressNodeFirewallDeny))
+			Expect(infwutils.CreateIngressNodeFirewall(testclient.Client, inf, timeout)).ShouldNot(HaveOccurred())
+			// TODO(mk): remove this sleep when we have a way to detect INF application is completed. Currently, we have no
+			// transition or pending state in the INF obj.
+			time.Sleep(200 * time.Millisecond)
```

Comment:
we not get nodeState object and check its status using eventually instead of sleep ?

Reply:
I am afraid theres a race there.
How do I know if the status has been updated after application of the configuration and doesnt just reflect the previous configurations status.

What Id like to observe is:
`Synchronized` -> `Pending` -> `Synchronized`

So, in this code, Id wait for pending state then wait for `Synchronized` status state. I would know then my configuration is applied. Not perfect since I could miss the `Pending` status state.

Need your input.